kyo-core/shared/src/main/scala/kyo/streams.scala

```diff
@@ -0,0 +1,154 @@
+                        case 0 =>
+                            ()
+                        case n =>
+                            Vars[TakeN].set(n - 1)
+                                .andThen(emit(v))
+                    }
+                }
+            }
+
+        def drop[T: Flat, S](n: Int)(v: T < (Streams[V] & S)): T < (Streams[V] & S) =
+            Vars[DropN].run(n) {
+                reemit(v) { v =>
+                    Vars[DropN].get.map {
+                        case 0 =>
+                            emit(v)
+                        case n =>
+                            Vars[DropN].set(n - 1)
+                    }
+                }
+            }
+
+        def filter[T: Flat, S](v: T < (Streams[V] & S))(
+            f: V => Boolean < S
+        ): T < (Streams[V] & S) =
+            reemit(v) { v =>
+                f(v).map {
+                    case false => ()
+                    case true  => emit(v)
+                }
+            }
+
+        def transform[T: Flat, V2: Flat, S, S2](v: T < (Streams[V] & S))(
+            f: V => V2 < (S & S2)
+        )(using Tag[Streams[V2]]): T < (Streams[V2] & S & S2) =
+            reemit(v) { v =>
+                f(v).map(Streams[V2].emit)
+            }
+
+        def collect[T: Flat, S, S2, V2: Flat](v: T < (Streams[V] & S))(
+            f: PartialFunction[V, Unit < (Streams[V2] & S2)]
+        ): T < (Streams[V2] & S & S2) =
+            reemit(v) { v =>
+                if f.isDefinedAt(v) then f(v)
+                else ()
+            }
+        end collect
+
+        private inline def reemit[T: Flat, S, S2, V2: Flat](v: T < (Streams[V] & S))(
+            inline f: V => Unit < (Streams[V2] & S2)
+        ): T < (Streams[V2] & S & S2) =
+            def loop[T2: Flat, S](v: T2 < (Streams[V] & S)): T2 < (Streams[V2] & S & S2) =
+                val handler = new Handler[Const[V], Streams[V], Streams[V2] & S2]:
+                    def resume[T3, U: Flat, S3](command: V, k: T3 => U < (Streams[V] & S3)) =
+                        f(command).andThen(loop(k(().asInstanceOf[T3])))
+                handle(handler, v)(using tag, Flat.derive)
+            end loop
+            loop[T, S](v)
+        end reemit
+
+        def runFold[T: Flat, S, A](v: T < (Streams[V] & S))(acc: A)(f: (A, V) => A)(
+            using Flat[V]
+        ): (A, T) < S =
+            def handler(acc: A): ResultHandler[Const[V], [T] =>> (A, T), Streams[V], S] =
+                new ResultHandler[Const[V], [T] =>> (A, T), Streams[V], S]:
+                    def pure[T](v: T) = (acc, v)
+                    def resume[T, U: Flat, S3](command: V, k: T => U < (Streams[V] & S3)) =
+                        handle(handler(f(acc, command)), k(().asInstanceOf[T]))
+            val a = handle(handler(acc), v)
+            a
+        end runFold
+
+        def runSeq[T: Flat, S](v: T < (Streams[V] & S))(
+            using Flat[V]
+        ): (Seq[V], T) < S =
+            def handler(acc: List[V])
```

Comment:
Perhaps worth adding a specific version of `runSeq` that only handles `Stream[V]` to avoid the somewhat awkward return type.

Reply:
I think that's a fundamental characteristic of the encoding since the the computation produces a value and a stream. We could have a `runSeqUnit` that discards the result of the computation but I'd prefer to keep only the canonical APIs for now.